File: smallest_subtree_with_all_the_deepest_nodes.go
// https://leetcode.com/problems/smallest-subtree-with-all-the-deepest-nodes/description/

/**
* Definition for a binary tree node.
	* type TreeNode struct {
	*     Val int
	*     Left *TreeNode
	*     Right *TreeNode
	* }
*/
func rec(root *TreeNode) (*TreeNode, int) {
	if root == nil {
		return nil, 0
	}

	if root.Left == nil && root.Right == nil {
		return root, 1
	}

	l, dl := rec(root.Left)
	r, dr := rec(root.Right)

	if dl > dr {
		return l, dl + 1
	} else if dl < dr {
		return r, dr + 1
	} else {
		return root, dl + 1
	}
}

func subtreeWithAllDeepest(root *TreeNode) *TreeNode {
	t, _ := rec(root)
	return t
}

File: algorithms/medium/smallest_subtree_with_all_the_deepest_nodes.kt
/**
 * Example: var ti = TreeNode(5) var v = ti.`val` Definition for a binary tree node. class
 * TreeNode(var `val`: Int) { var left: TreeNode? = null var right: TreeNode? = null }
 */
class Solution {
  fun rec(root: TreeNode?): Pair<TreeNode?, Int> {
    if (root == null) {
      return Pair(null, 0)
    }

    if (root?.left == null && root?.right == null) {
      return Pair(root, 1)
    }

    val (l, dl) = rec(root?.left)
    val (r, dr) = rec(root?.right)

    if (dl == dr) {
      return Pair(root, dl + 1)
    }

    if (dl > dr) {
      return Pair(l, dl + 1)
    }

    return Pair(r, dr + 1)
  }

  fun subtreeWithAllDeepest(root: TreeNode?): TreeNode? {
    val (node, _) = rec(root)
    return node!!
  }
}
